### PR TITLE
v3: support for aggregation and sorting of TEXT

### PR DESCRIPTION
**Pull request by sougou**
This change adds support for aggregating and sorting of columns
that may need to follow collation rules.

The feature depends on mysql's weight_string function
that returns a lexically comparable value of any text column.
If a group by or order by uses a text column like varchar, then
v3 will additionally request the weight_string versions of such
columns and perform ordering and aggregation using those values
instead.

The identification of a text column is currently based on a new
"columns" field in the vschema that allows you to specify the
type of each column. This part was implemented in a previous PR.

Two approaches were possible:
1. Request additional weight_string values at the time of push-down.
2. Rewire the primitives to request and use the additional weight_string
   values during the Wireup phase.
I went with option 2 because it minimizes overall impact of the
code, which will allow us to yank this behavior out when we
implement collation aware sorting in vtgate. Also, since all
the weight_string columns get added at the end, we only have
to truncate the rows before returning what's needed.

The following changes were made:
* sqltypes: result truncate functionality.
* engine: Add TruncateColumnCount field that can be used to truncate
  a result if needed.
* planbuilder: Change Wireup to check and request weight_strings.
  If weight_string was requested, set the row truncation to make
  sure that the weight_string values don't get passed on.

### CODE REVIEW

**Comment by dumbunny** [comment]
LGTM

File: go/vt/vtgate/planbuilder/route.go [hunk 1]
```diff
@@ -57,16 +57,22 @@ type route struct {
...
 func newRoute(stmt sqlparser.SelectStatement, eroute *engine.Route, condition sqlparser.Expr, vschema VSchema) *route {
 	rb := &route{
-		Select:    stmt,
-		order:     1,
-		condition: condition,
-		ERoute:    eroute,
+		Select:        stmt,
+		order:         1,
+		condition:     condition,
+		weightStrings: make(map[*resultColumn]int),
```
Comment (demmer) [inline go/vt/vtgate/planbuilder/route.go]: It seems cleaner and I suspect more efficient on the gc to only create `weightStrings` when needed rather than always create the empty array.

File: go/vt/vtgate/vindexes/lookup_hash.go [hunk 2]
```diff
@@ -146,7 +144,6 @@ type LookupHashUnique struct {
 
 // NewLookupHashUnique creates a LookupHashUnique vindex.
 func NewLookupHashUnique(name string, m map[string]string) (Vindex, error) {
-	log.Warningf("LookupHashUnique index (%q) it's being deprecated. Please use LookupUnique", name)
```
Comment (demmer) [inline go/vt/vtgate/vindexes/lookup_hash.go]: I assume removing these is intentional (which I support) but that seems orthogonal to these changes.

File: go/vt/vtgate/planbuilder/route.go (hunk 1 above)
Comment (sougou) [reply go/vt/vtgate/planbuilder/route.go]: Unfortunately, this is a map. There's no autocreation support for it like slices.
Deferring the creation will introduce conditionals in multiple places (nil-check->make).

Besides, this is planbuilder time, which should be low QPS ideally.

File: go/vt/vtgate/vindexes/lookup_hash.go (hunk 2 above)
Comment (sougou) [reply go/vt/vtgate/vindexes/lookup_hash.go]: Yeah. TThere was too much spam in tests.

File: go/sqltypes/result.go [hunk 3]
```diff
@@ -98,6 +98,41 @@ func CopyRow(r []Value) []Value {
 	return out
 }
 
+// Truncate returns a new Result with all the rows truncated
```
[the remark this answers is not in the document]
Comment (sougou) [reply go/sqltypes/result.go]: Done.

**Comment by indera-shsp** [comment]
@sougou IS this a known bug?

```
select id, companyName As test from companyProfile where id = 3 order by companyName;
+----+-------------+
| id | test        |
+----+-------------+
|  3 | SharpSpring |
+----+-------------+
1 row in set (0.09 sec)


select id, companyName As test from companyProfile where id in(3) order by companyName;
ERROR 1064 (42000): vtgate: http://vtgate-staging-external-867fc76d86-p4w9k:15001/: target: sharpspring_app.-80.master, used tablet: staging-978852100 (staging-sharpspring-app-x-80x-replica-0.vttablet.default.svc.cluster.local): vttablet: rpc error: code = InvalidArgument desc = You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near 'test) from companyProfile where 1 != 1' at line 1 (errno 1064) (sqlstate 42000) (CallerID: andrei.sura): Sql: "selec
```
